crypto/src/key/hdkd/derivation_path.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::DerivationError;
+use crate::key::hdkd::DerivationError::*;
+use core::fmt;
+use std::fmt::Formatter;
+use std::slice::Iter;
+use std::str::FromStr;
+
+const PREFIX: &str = "m";
+const SEPARATOR: &str = "/";
+const HARDENED_BIT: u32 = 0x80000000;
+
+/// BIP32-like child numbers. Currently we only support hardened derivations
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub enum ChildNumber {
+    Hardened(u32),
+}
+
+impl ChildNumber {
+    /// Return a hardened child number
+    pub fn hardened(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            Ok(ChildNumber::Hardened(index))
+        } else {
+            Err(InvalidChildNumber(index))
+        }
```

Comment:
Did you mean checking these errors at run time?

Using a `u31` type will push the checking the result in to that type. I took the approach that @TheQuantumPhysicist suggested, wrapping a private enum in a struct:

```rust
#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
pub struct ChildNumber(DerivationType);

#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
enum DerivationType {
    Hardened(u32),
}
```

So this was it would be impossible to create a wrong kind of ChildNumber i.e. with the internal `u32` having the MSB set to 1. It is not exactly what you were referring to but let me know if this approach is better.

Reply:
Just to clarify, there's also a NonZeroU32, NonZeroU64, etc in rust that solve a similar problem. The benefit here is to ensure that invariants are held across many functions and environments. For example, if function A calls B, and B calls C, ideally you'd have to check the u32 parameter in every call, which is cumbersome. But then if you enforce using U31, then you're guaranteed to do this only once by extracting the logic outside of these functions.